src/pybind/mgr/volumes/subvolume.py

```diff
@@ -228,11 +160,62 @@ def rmtree(root_path):
 
                 d = self.fs.readdir(dir_handle)
             self.fs.closedir(dir_handle)
-
             self.fs.rmdir(root_path)
 
-        rmtree(trashed_subvolume)
+        trashpath = spec.make_trash_path()
+        rmtree(trashpath)
+
+    def list_subvolumes(self, spec):
+        """
+        Returns a list of subvolumes within a subvolume group
+        """
+        grouppath = spec.make_group_path()
+        return self._list_dir(grouppath)
+
+    def get_subvolume_path(self, spec):
+        path = spec.make_fs_path()
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            return None
+        except cephfs.Error as e:
+            raise VolumeException(e.args[0]. e.args[1])
+        return path
 
+    ### Group operations...
+
+    def create_group(self, spec, mode=0o755):
+        path = spec.make_group_path()
+        self._mkdir_p(path, mode)
+
+    def remove_group(self, spec, force):
+        path = spec.make_group_path()
+        try:
+            self.fs.rmdir(path)
+        except cephfs.ObjectNotFound:
+            if not force:
+                raise VolumeException(-errno.ENOENT, "Subvolume group '{0}' not found".format(spec.get_group_id()))
+        except cephfs.Error as e:
+            raise VolumeException(e.args[0], e.args[1])
+
+    def list_groups(self):
+        """
+        Returns a list of subvolume groups within a FS volume/filesystem
+        """
+        spec = SubvolumeSpec("", "")
+        path = spec.get_group_dir()
+        return self._list_dir(path)
+
+    def get_group_path(self, spec, create_default=False):
+        path = spec.make_group_path()
+        try:
+            self.fs.stat(path)
+        except cephfs.ObjectNotFound:
+            if create_default and spec.is_default_group():
+                self._mkdir_p(path)
+                return path
+            return None
```

Comment:
caller often raise an exception
```suggestion
            raise VolumeException(
                        -errno.ENOENT, "Subvolume group '{0}' not found".format(groupname))
```
does it make sense to move this into `get_group_path`?

Reply:
ditto as per `get_subvolume_path()`